File: packages/sdk/src/providers/openai.ts
import type { WrapContext } from '../config.js';

import { wrapCreate } from './shared.js';
import type { ProviderAdapter, ProviderHostLogger } from './types.js';

/** Keys forwarded into `modelConfiguration` from the OpenAI request params. */
export const OPENAI_CONFIG_KEYS: readonly string[] = [
  'temperature',
  'top_p',
  'max_tokens',
  'frequency_penalty',
  'presence_penalty',
];

/** Keys retained from the OpenAI response when building the output snapshot. */
export const OPENAI_OUTPUT_KEYS: readonly string[] = ['choices', 'usage', 'model', 'id'];

type CreateFn = (...args: unknown[]) => Promise<unknown>;

interface OpenAiClient extends Record<string, unknown> {
  chat: { completions: { create: CreateFn } };
}

export const openaiAdapter: ProviderAdapter = {
  providerId: 'openai',

  detect(client: object): boolean {
    const c = client as Record<string, unknown>;
    const chat = c['chat'] as Record<string, unknown> | undefined;
    const completions = chat?.['completions'] as Record<string, unknown> | undefined;
    return typeof completions?.['create'] === 'function';
  },

  wrap(audit: ProviderHostLogger, client: object, context: WrapContext): void {
    const openai = client as OpenAiClient;
    const completions = openai.chat.completions;
    const originalCreate = completions.create.bind(completions);
    completions.create = wrapCreate(audit, originalCreate, context, {
      providerId: 'openai',
      configKeys: OPENAI_CONFIG_KEYS,
      outputKeys: OPENAI_OUTPUT_KEYS,
      // OpenAI doesn't expose a snapshot suffix the way Anthropic does;
      // reuse the model id so AuditLogEntryInputSchema's modelVersion.min(1)
      // constraint is always satisfied.
      buildModelInfo: (params) => {
        const model = String(params['model'] ?? 'unknown');
        return { modelName: model, modelVersion: model };
      },
      buildInput: (params) => ({ messages: params['messages'] }),
    });
  },
};

File: python/src/vouchrail/providers/openai.py
"""OpenAI provider adapter (sync wrap_openai)."""

from __future__ import annotations

import inspect
from typing import Any

from ._shared import WrapCreateOptions, build_sync_wrapper
from .base import ProviderHostLogger, WrapContext

OPENAI_CONFIG_KEYS: tuple[str, ...] = (
    "temperature",
    "top_p",
    "max_tokens",
    "frequency_penalty",
    "presence_penalty",
)

OPENAI_OUTPUT_KEYS: tuple[str, ...] = ("choices", "usage", "model", "id")


def _openai_wrap_options() -> WrapCreateOptions:
    def model_info(params: dict[str, Any]) -> tuple[str, str]:
        model = str(params.get("model") or "unknown")
        return model, model

    def input_snapshot(params: dict[str, Any]) -> Any:
        return {"messages": params.get("messages")}

    return WrapCreateOptions(
        provider_id="openai",
        config_keys=OPENAI_CONFIG_KEYS,
        output_keys=OPENAI_OUTPUT_KEYS,
        build_model_info=model_info,
        build_input=input_snapshot,
    )


class OpenAiAdapter:
    provider_id = "openai"

    def detect(self, client: Any) -> bool:
        chat = getattr(client, "chat", None)
        if chat is None:
            return False
        completions = getattr(chat, "completions", None)
        if completions is None:
            return False
        create = getattr(completions, "create", None)
        if not callable(create):
            return False
        return not inspect.iscoroutinefunction(create)

    def wrap(self, audit: ProviderHostLogger, client: Any, context: WrapContext) -> None:
        completions = client.chat.completions
        completions.create = build_sync_wrapper(
            audit, completions.create, context, _openai_wrap_options(),
        )


openai_adapter = OpenAiAdapter()
